File: app/src/main/java/com/haishinkit/rtmp/messages/RTMPAVCVideoMessage.java
package com.haishinkit.rtmp.messages;

import com.haishinkit.flv.VideoCodec;
import com.haishinkit.rtmp.RTMPSocket;

import java.nio.ByteBuffer;

public final class RTMPAVCVideoMessage extends RTMPVideoMessage {
    private byte packetType = 0;
    private int compositeTime = 0;

    public RTMPAVCVideoMessage() {
        super();
        setCodec(VideoCodec.AVC.rawValue());
    }

    public byte getPacketType() {
        return packetType;
    }

    public RTMPAVCVideoMessage setPacketType(final byte packetType) {
        this.packetType = packetType;
        return this;
    }

    public int getCompositeTime() {
        return compositeTime;
    }

    public RTMPAVCVideoMessage setCompositeTime(final byte compositeTime) {
        this.compositeTime = compositeTime;
        return this;
    }

    @Override
    public ByteBuffer encode(final RTMPSocket socket) {
        if (socket == null) {
            throw new IllegalArgumentException();
        }
        int length = getPayload() == null ? 0 : getPayload().limit();
        ByteBuffer buffer = ByteBuffer.allocate(5 + length);
        buffer.put((byte)(getFrame() << 4 | getCodec()));
        buffer.put(getPacketType());
        buffer.put(new byte[]{0x00, 0x00, 0x00});
        if (0 < length) {
            buffer.put(getPayload());
        }
        return buffer;
    }

    @Override
    public RTMPMessage decode(final ByteBuffer buffer) {
        if (buffer == null) {
            throw new IllegalArgumentException();
        }
        buffer.position(buffer.position() + getLength());
        return this;
    }
}

File: haishinkit/src/main/java/com/haishinkit/rtmp/messages/RtmpAvcVideoMessage.kt
package com.haishinkit.rtmp.messages

import com.haishinkit.flv.FlvAvcPacketType
import com.haishinkit.flv.FlvVideoCodec
import com.haishinkit.iso.AvcFormatUtils
import java.nio.ByteBuffer

internal class RtmpAvcVideoMessage : RtmpVideoMessage() {
    override var codec = FlvVideoCodec.AVC
    var packetType: Byte = 0
    var compositeTime = 0
    override var length: Int
        get() = 5 + (data?.limit() ?: 0)
        set(value) { super.length = value }
    override var payload: ByteBuffer = EMPTY_BYTE_BUFFER
        get() {
            if (field.capacity() < length) {
                field = ByteBuffer.allocate(5 + (data?.capacity() ?: 0))
            } else {
                field.clear()
            }
            return field
        }

    override fun encode(buffer: ByteBuffer): RtmpMessage {
        buffer.put((frame.toInt() shl 4 or codec.toInt()).toByte())
        buffer.put(packetType)
        buffer.put((compositeTime shr 16).toByte()).put((compositeTime shr 8).toByte()).put(compositeTime.toByte())
        data?.let {
            when (packetType) {
                FlvAvcPacketType.NAL -> {
                    AvcFormatUtils.put(it, buffer)
                }
                else -> {
                    buffer.put(it)
                }
            }
        }
        return this
    }

    override fun decode(buffer: ByteBuffer): RtmpMessage {
        buffer.position(buffer.position() + length)
        return this
    }
}
